### PR TITLE
generic: cleanup duplicate code in handling unexp request

### PR DESCRIPTION
**Pull request by yfguo**
## Pull Request Description

The copying from unexp buffer to user buffer is duplicated in irecv, imrecv and completion handling. This PR merges the code, so it is easier to follow.

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by yfguo** [comment]
test:mpich/ch4/most

File: src/mpid/ch4/generic/am/mpidig_am_recv_utils.h [hunk 1]
```diff
@@ -74,7 +74,7 @@ MPL_STATIC_INLINE_PREFIX void MPIDIG_recv_finish(MPIR_Request * rreq)
 /* Transport-specific data copy, such as RDMA, need explicit iov pointers.
  * Providing helper routine keeps the internal of MPIDIG_rreq_async_t here.
  */
-MPL_STATIC_INLINE_PREFIX void MPIDIG_convert_datatype(MPIR_Request * rreq);
+MPL_STATIC_INLINE_PREFIX void MPIDIG_convert_datatype(MPIR_Request * req);
```
[the remark this answers is not in the document]
Comment (yfguo) [reply src/mpid/ch4/generic/am/mpidig_am_recv_utils.h:77]: No, I will revert this. Was doing a search and replace.

File: src/mpid/ch4/generic/am/mpidig_am_recv_utils.h [hunk 2]
```diff
@@ -274,42 +274,112 @@ MPL_STATIC_INLINE_PREFIX int MPIDIG_recv_copy_seg(void *payload, MPI_Aint payloa
...
+/* utility function for copying data from the unexp buffer into user buffer. This avoid the dup code
+ * for handling unexp message in am_do_irecv, am_do_imrecv, and completion. Specifically involves
+ * the followin cases:
+ * 1. am_do_irecv: unexp req matched when posting Irecv
+ * 2. am_do_imrecv: unexp req matched when posting Imrecv
+ * 3. completion: unexp req matched with Irecv/Imrecv enqueued by someone else (WorkQ model)
+ * */
+MPL_STATIC_INLINE_PREFIX int MPIDIG_recv_from_unexp_req(MPIR_Request * req, void *user_buf,
+                                                        MPI_Datatype user_datatype,
+                                                        MPI_Aint user_count)
```
[the remark this answers is not in the document]
Comment (yfguo) [reply src/mpid/ch4/generic/am/mpidig_am_recv_utils.h:286]: Sure. Let me rename it and put it in `generic/am/mpidig_am_recv.h` then as it is mostly used there.

**Comment by yfguo** [comment]
test:mpich/ch4/most

**Comment by yfguo** [comment]
Test results were clean. Pushed to fix warning.

File: src/mpid/ch4/generic/am/mpidig_am_recv.h [hunk 3]
```diff
@@ -109,7 +109,8 @@ MPL_STATIC_INLINE_PREFIX int MPIDIG_handle_unexpected(void *buf, MPI_Aint count,
     rreq->status.MPI_SOURCE = MPIDIG_REQUEST(rreq, rank);
     rreq->status.MPI_TAG = MPIDIG_REQUEST(rreq, tag);
 
-    MPIDIG_copy_from_unexp_req(rreq, buf, datatype, count);
+    mpi_errno = MPIDIG_copy_from_unexp_req(rreq, buf, datatype, count);
+    MPIR_ERR_CHECK(mpi_errno);
```
Comment (hzhou) [inline src/mpid/ch4/generic/am/mpidig_am_recv.h:143]: I am on the 2nd commit. This change (adding return) belong to the previous commit.

Comment (yfguo) [reply src/mpid/ch4/generic/am/mpidig_am_recv.h:143]: Fixed.

**Comment by yfguo** [comment]
Last warning test seems dead on Jenkins. Retrigger it.

**Comment by hzhou** [comment]
test:mpich/ch4/most

**Review by hzhou — APPROVED** [review]
LGTM

**Pull request merged**